### PR TITLE
fix(amf): Fix for multisession service request

### PR DESCRIPTION
**Pull request by moinuddin1980**
fix(amf): Fix for multisession service request

Signed-off-by: Moinuddin Khan <moinuddin.khan@wavelabs.ai>

## Summary
Following fixes are included:
  1. AMF is not sending Service Accept. As per the new changes in sessiond, it expects session id information during service request procedures. Due to which sessiond was not processing the request.
   2. Service accept to be sent when all PDU session response (w.r.t. service request) is received from sessiond. 
   3. ICS + Service request to be send when all PDUs are accepted by sessiond. 
   4. In case of Service request using Uplink Nas Transport only Service Accept needs to be send on Downlink Nas transport (and not with ICS).
   
## Test Plan
Tested the TC2d using TVM setup which is passing with proper Service Accept and ICS Req messages
[SA_TC2d_Verify_Paging_with_Multi_PDU_session_220127_044502-1.zip](https://github.com/magma/magma/files/7952837/SA_TC2d_Verify_Paging_with_Multi_PDU_session_220127_044502-1.zip)
 

## Additional Information
Addresses the following tickets
#11089 
#11132  
- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_recv.cpp [hunk 1]
```diff
@@ -149,39 +237,83 @@ int amf_handle_service_request(
...
+            ue_context->pdu_session_reactivation_status = pdu_session_status;
+            ue_context->service_response_timer.id = amf_app_start_timer(
+                    SERVICE_RESEPONE_TIMER_EXPIRY_MSECS,
+                    TIMER_REPEAT_ONCE,
+                    service_rsponse_timer_handler,
+                    ue_context->amf_ue_ngap_id);
+
+            OAILOG_INFO(
+                    LOG_AMF_APP,
+                    "Starting service respone Timer for UE ID: " AMF_UE_NGAP_ID_FMT
```
[the remark this answers is not in the document]
Comment (panyogesh) [reply lte/gateway/c/core/oai/tasks/amf/amf_recv.cpp]: > fix(amf):Fixed 1.AMF not sending Service Accept, 2.multi PDU session paging issue
> 
> Signed-off-by: Moinuddin Khan [moinuddin.khan@wavelabs.ai](mailto:moinuddin.khan@wavelabs.ai)
> 
> ## Summary
> Fixed the following: In paging scenario upon getting Service Request from UE, AMF is not responding with Service Accept towards UE.
> 
> ## Test Plan
> Tested the TC2d using TVM setup which is passing with proper Service Accept and ICS Req messages [SA_TC2d_Verify_Paging_with_Multi_PDU_session_220127_044502-1.zip](https://github.com/magma/magma/files/7952837/SA_TC2d_Verify_Paging_with_Multi_PDU_session_220127_044502-1.zip)
> 
> ## Additional Information
> Addresses the following tickets #11089 #11132
> 
> * [ ]  This change is backwards-breaking

1. Can you please change the heading at put one liner summary like :                             
  "feat(amf): Multisession paging support" or "fix(amf): Multisession paging changes"

2. In Summary u can list the fixes (1. AMF not sending service accept: followed a brief description)

3. In "Test Plan" : Please list the scenarios tested if possibly mapping to fixes.

**Comment by panyogesh** [comment]
Can you please compare the pcaps (Have attached a smaple)
Pcap-Comp

**Comment by panyogesh** [comment]
Can we confirm that there is one PDUSessionResourceSetupListCxtReq and all other items like PDUSessionResourceSetupItemCtxReq under it.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 2]
```diff
@@ -618,40 +618,88 @@ int amf_app_handle_pdu_session_response(
...
+  }
+  else
+  {
+      smf_ctx->pdu_session_state = ACTIVE;
+      bool all_pdu_active        = true;
+
+      for (uint16_t session_id = 1; session_id < (sizeof(session_id) * 8);
+              session_id++) {
+          if (ue_context->pdu_session_status & (1 << session_id))
+          {
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Can we  make use of "smf_ctx->pdu_session_state". We can mark it inactive when paging notification is send from AMF to SMF.

File: lte/gateway/c/core/oai/tasks/amf/amf_recv.cpp [hunk 3]
```diff
@@ -149,39 +237,83 @@ int amf_handle_service_request(
...
+                    }
+                }
+            }
+
+            /* Start Service response timer */
+            ue_context->pending_service_response = true;
+            ue_context->pdu_session_status = pdu_session_status;
+            ue_context->pdu_session_reactivation_status = pdu_session_status;
+            ue_context->service_response_timer.id = amf_app_start_timer(
+                    SERVICE_RESEPONE_TIMER_EXPIRY_MSECS,
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/amf_recv.cpp]: Timer in UE T3517 can be helpful to figure out if service request needs to be aborted.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 2 above)
Comment (moinuddin1980) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Taken care

File: lte/gateway/c/core/oai/tasks/amf/amf_recv.cpp (hunk 3 above)
Comment (moinuddin1980) [reply lte/gateway/c/core/oai/tasks/amf/amf_recv.cpp]: Taken care

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 4]
```diff
@@ -623,40 +623,81 @@ int amf_app_handle_pdu_session_response(
...
+      OAILOG_DEBUG(
+          LOG_AMF_APP,
+          "Sending message to gNB for PDUSessionResourceSetupRequest "
+          "**n_active_pdus=%d **\n",
+          smf_ctx->n_active_pdus);
+
+      amf_smf_msg.pdu_session_id = pdu_session_resp->pdu_session_id;
+      /*Execute PDU establishement accept from AMF to gnodeb */
+      pdu_state_handle_message(
+          // ue_context->mm_state, STATE_PDU_SESSION_ESTABLISHMENT_ACCEPT,
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Please remove commented lines if not required.

Comment (moinuddin1980) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Done

File: lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp [hunk 5]
```diff
@@ -408,4 +408,86 @@ bool get_ue_id_from_imsi(
...
+   * message has uplink data status(4 bytess)
+   * and PDU session status(4 bytes) after TMSI*/
+  ue_tmsi = htonl(ue_tmsi);
+  memcpy(
+      &(initial_ue_message.nas->data[nas_msg_length - sizeof(tmsi_t) - 8]),
+      &(ue_tmsi), sizeof(tmsi_t));
+  initial_ue_message.sctp_assoc_id               = sctp_assoc_id;
+  initial_ue_message.gnb_ue_ngap_id              = gnb_ue_ngap_id;
+  initial_ue_message.gnb_id                      = gnb_id;
+  initial_ue_message.m5g_rrc_establishment_cause = M5G_MO_SIGNALLING;
```
[the remark this answers is not in the document]
Comment (moinuddin1980) [reply lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp]: Done

File: lte/gateway/c/core/oai/tasks/amf/amf_app_defs.h [hunk 6]
```diff
@@ -53,4 +53,9 @@ void amf_app_handle_initial_context_setup_rsp(
     amf_app_desc_t* amf_app_desc_p,
     itti_amf_app_initial_context_setup_rsp_t* initial_context_rsp);
 int amf_send_n11_update_location_req(amf_ue_ngap_id_t ue_id);
+// Handling context release complete
+void amf_app_handle_ngap_ue_context_release_complete(
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_defs.h]: Is this being used.

**Comment by panyogesh** [comment]
Please use a new commit-id for reviews so that delta changes are visible.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_defs.h (hunk 6 above)
Comment (moinuddin1980) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_defs.h]: Removed it. Done

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**